### ISSUE TITLE
CS0452 and CS0718 error sometimes reported on unintuitive locations

### ISSUE THREAD

**Issue by Rekkonnect** [dotnet/roslyn#65515]
The [CS0452](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0452) and [CS0718](https://docs.microsoft.com/en-us/dotnet/csharp/misc/cs0718) errors are inconsistently reported on either the type arguments themselves, or the symbol of the types.

Specifically, they are reported on the symbol *using* the type `G` with the erroneous type arguments when it's:
- a type that inherits `G`
- a method returning `G`
- a method parameter of type `G`
- a field of type `G`
- a property of type `G`

Wheras, the errors are reported on the erroneous type arguments themselves when the symbol is:
- a local of type `G`

Following #61831, I'm directly in favor of keeping the behavior that already applies to locals, which offers the following advantages:
- error locality, directing the user closer towards the real problem
- less errors on the same node (currently every type argument's error stacks on top of other type arguments' errors)
- consistency with other errors
- consistency across symbol types that may emit the same errors

## Example

[SharpLab repro](https://sharplab.io/#v2:EYLgtghglgdgNAFxANwKYCcoDMCeAfAAQGYACAgJhIBEMplYBzEkEgIQgGdUAeAZQQgIoAYwCMcMqIAMEgDYB7GAwB8AWABQAbw0BfDRuJlK7LtwAqEgKoSAamvUlHJAO4ALDKhJnmRjU5fu6J6WPhR+Tm4eJDah5Bra6nrqBqQUJJYcjCao8eGOAA6YyIKe2XwCQmISBNJyiiokWFCosgAmANx5ZKRl/IIi4pIyJApKyiQACujy+RgIOCSaJAyoCO0kXGskSV2GvRUD1bUj9eMAsquu8q0AFPv9VUN1YySwwvJgjACUXQn+/vdKoMasNRg0ghwAK6yBAkAC8JBg0NknQc/0cBAA7CQIdC1l0kjt1IYagA2IwkPpA3LJRJAA)

*For some reason the repro on SharpLab does not report any errors on the `incoming` parameter; this might be some other bug.*

**Current Behavior**:
```csharp
public class Deriving : Base<Static1, string, long>
             ~~~~~~~~
{
}

public class Base<T, U, V>
    where T : class
    where U : class
    where V : class
{
}

public class UsingBase
{
    private Base<Static1, string, long> field;
                                        ~~~~~
    public Base<Static1, string, long> Property { get; set; }
                                       ~~~~~~~~

    public Base<Static1, string, long> Method(Base<Static1, string, long> incoming)
                                       ~~~~~~                             ~~~~~~~~
    {
        Base<Static1, string, long> result = null;
             ~~~~~~~          ~~~~
        return result;
    }
}

public static class Static1
{

}
```

**Expected Behavior**:
```csharp
public class Deriving : Base<Static1, string, long>
                             ~~~~~~~          ~~~~
{
}

public class Base<T, U, V>
    where T : class
    where U : class
    where V : class
{
}

public class UsingBase
{
    private Base<Static1, string, long> field;
                 ~~~~~~~          ~~~~
    public Base<Static1, string, long> Property { get; set; }
                ~~~~~~~          ~~~~

    public Base<Static1, string, long> Method(Base<Static1, string, long> incoming)
                ~~~~~~~          ~~~~              ~~~~~~~          ~~~~
    {
        Base<Static1, string, long> result = null;
             ~~~~~~~          ~~~~
        return result;
    }
}

public static class Static1
{

}
```

**Comment by jcouv** [comment]
I agree the diagnostic locations (directly on faulty type arguments) would be more helpful.
FWIW, the two errors on parameter `incoming` are properly reported in VS. Not sure why it's not showing up in sharplab...

**Comment by Rekkonnect** [comment]
That's weird; using VS 17.5.0 Preview 1.0 I found the errors be reported on `incoming` instead of its type arguments.

**Comment by Rekkonnect** [comment]
I have modified the thread info and the repro to include aliases, which are also found to behave the same. There arises another concern now:
- **Is it worthy of reporting the errors about the incorrect type arguments on both the alias declaration **and** the usages of the alias type?**
By migrating all error locations to the type arguments themselves, this is directly avoided by filtering by whether the using type is an alias type.

**Comment by Rekkonnect** [comment]
With a little bit of digging into the code what stood out for me is that [`ConstraintsHelper.CheckConstraintsArgs`](https://github.com/dotnet/roslyn/blob/00f49cc4587f982524e726a801e946ddf7cf42ed/src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs#L491-L519) only contains a location, without a reference to a syntax node, nor providing the option for one. The reason why in expressions and locals the errors are reported on the erroneous type arugments is because they use the type argument syntax list to get the position of the diagnostic, ignoring the provided location in the `args` parameter in the methods inside `ConstraintsHelper`.

Based on the above, to move this forward, this code seems like it should be refactored and pass down the syntax references to the declarations, if any, otherwise defaulting back to the location argument.

**Comment by CyrusNajmabadi** [comment]
Dupe of https://github.com/dotnet/roslyn/issues/20172
